keps/sig-scheduling/2891-simplified-config/README.md

```diff
@@ -0,0 +1,553 @@
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP defines a simplified field for end-users to configure Scheduler plugins
+which use multiple extension points easily. It is based on a new extension point that
+plugin developers can implement in their plugins. This extension point will automatically
+register a plugin for all (or some) of the scheduling cycle extension points that the
+plugin implements. Thus simplifying the configuration required for cluster administrators
+while preserving the development benefits of the scheduling framework.
+
+
+## Motivation
+
+The [Scheduling Framework](https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/624-scheduling-framework) was designed with
+developers of custom schedulers in mind to make the scheduler more extendable. This design
+provided several extension points used by developers to implement their custom plugins into
+the scheduling cycle. Some extension points, like `PreFilter` and `PreScore`, are only
+used for data processing or other internal functions not directly relevant to the end user.
+
+However these "behind-the-scenes" endpoints must still be configured in the `KubeSchedulerConfiguration`
+provided by the user for every plugin that uses them. For the average user this is an unnecessary
+and confusing step -- they only care that the plugin works, not how it works, and so
+internal processing steps are usually irrelevant to them.
+
+Therefore, a way for users to simply toggle all of a plugin's extension points
+(or a subset of the extension points as defined necessary by the plugin developer)
+would remove complexity and the potential for misconfiguration.
+
+### Goals
+
+* Provide an option for simplified plugin config from the end-user's perspective
+* Enable this config in all in-tree default scheduler plugins
+
+### Non-Goals
+
+* Refactor or introduce any breaking changes to the scheduling framework
+
+## Proposal
+
+A new extension point, `MultiPoint` (final name TBD) will be made available to
+users and developers which handles plugin registration for multiple extension
+points simultaneously.
+
+From the user's perspective, this will be a new (optional) field in the scheduler's
+config to enable or disable plugins across multiple extension points.
+
+
+### User Stories (Optional)
```

Comment:
One important point here is that this multipoint approach doesn't solve the problem of refactoring plugins to add new extension points since we will continue to support the fine grained configuration. We can only do this with a new CC version. Perhaps worth pointing that out somewhere?

Reply:
I think we should consider fusing PreScore + Score, PreFilter + Filter and PreBind+Bind from the config perspective. Like if Filter is specified, we check if both PreFilter and/or Filter extensions are implemented.

Reply:
> One important point here is that this multipoint approach doesn't solve the problem of refactoring plugins to add new extension points since we will continue to support the fine grained configuration

Could you explain this a bit more? If a plugin is refactored to implement a new extension point, this multipoint approach will automatically register it in the new extension point as well.

Reply:
The multipoint will handle it, but we are still supporting the original configuration where admins could explicitly list the set of extensions, and so if we refactor a plugin that initially implemented only Score to PreScore+Score, they will be surprised that their configuration which included enabling only the Score extension is no longer sufficient.

Reply:
we should take a position on recommending the new configuration to administrators since it addresses problems like that. If they are still using the old one while an alternative exists, then I think known issues like that can be expected.

I'm not opposed to changing some of the old config like you suggested, but I think it's not relevant to multipoint since it is solving that problem for itself

Reply:
I don't think this proposal is totally irrelevant. The core of the MultiPoint proposal is to preserve the old configuration approach, right? If we didn't want admins to use the old approach, we wouldn't introduce MultiPoint, we would rather just force it all the time as I initially suggested in https://github.com/kubernetes/kubernetes/issues/102303. One of the motivations for https://github.com/kubernetes/kubernetes/issues/102303 was actually allowing to re-factor plugins, see quote below 

> If we decide to re-factor a plugin to add more extension points, this becomes a backward incompatible change since adding or removing an extension point may break existing configurations (see this example Scheduler(NodeVolumeLimits plugin): filter volumes of new pod in prefilter phase #102235 (comment)).

Making a recommendation to only use MultiPoint isn't sufficient to introduce changes that break the old config approach since it is still officially supported. 

In any case, I think such refactorings are usually limited to doing pre-calculations, hence the suggestion to fuse all Pre* extensions just like we did with Normalize and I feel this should be part of this proposal (simplifying plugin config), but I am not going to veto the proposal without doing this if others think it is not the time to pursue it or significantly increases the scope of this enhancement. @Huang-Wei @alculquicondor  wdyt?

Reply:
There is also one semantic problem:
PreFilter can be used to mark a Pod Unschedulable at the cluster level (without the need to check any Nodes). So in a way PreFilter might exist on its own.

Maybe PreScore doesn't make sense on its own, but it's kind of awkward if PreScore is merged with Score but PreFilter is not merged with Filter.

PostFilter is supposed to be separate from Filter in any case. This is not in consideration for merging.

Reply:
I think enlarging the scope is fine, shouldn't ideally be a blocker. I do think it is something we should seriously consider, may be in a v1beta4 :)

Reply:
> PreFilter can be used to mark a Pod Unschedulable at the cluster level (without the need to check any Nodes). So in a way PreFilter might exist on its own.

That is fine, any plugin listed under "Filter" can implement either both or one of them.